src/hotspot/share/services/memMapPrinter.cpp

```diff
@@ -0,0 +1,303 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+
+#include "logging/logAsyncWriter.hpp"
+#include "gc/shared/collectedHeap.hpp"
+#include "memory/allocation.hpp"
+#include "memory/universe.hpp"
+#include "runtime/nonJavaThread.hpp"
+#include "runtime/osThread.hpp"
+#include "runtime/thread.hpp"
+#include "runtime/threadSMR.hpp"
+#include "runtime/vmThread.hpp"
+#include "services/memMapPrinter.hpp"
+#include "nmt/memTracker.hpp"
+#include "nmt/virtualMemoryTracker.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/ostream.hpp"
+
+// Note: throughout this code we will use the term "VMA" for OS system level memory mapping
+
+/// NMT mechanics
+
+// Short, clear, descriptive names for all possible markers. Note that we only expect to see
+// those that have been used with mmap. Others I leave at nullptr.
+#define NMTFLAGS_DO(f) \
+  /* flag, short, description */ \
+  f(mtGCCardSet,      "CARDTBL", "GC Card table") \
+  f(mtClassShared,    "CDS", "CDS archives") \
+  f(mtClass,          "CLASS", "Class Space") \
+  f(mtCode,           "CODE", "Code Heap") \
+  f(mtGC,             "GC", "GC support data (e.g. bitmaps)") \
+  f(mtInternal,       "INTERN", "Internal") \
+  f(mtJavaHeap,       "JAVAHEAP", "Java Heap") \
+  f(mtOther,          "JDK", "allocated by JDK libraries other than VM") \
+  f(mtMetaspace,      "META", "Metaspace nodes (non-class)") \
+  f(mtSafepoint,      "POLL", "Polling pages") \
+  f(mtThreadStack,    "STACK", "(known) Thread Stack")
+  //end
+
+static const char* get_shortname_for_nmt_flag(MEMFLAGS f) {
+#define DO(flag, shortname, text) if (flag == f) return shortname;
+  NMTFLAGS_DO(DO)
+#undef DO
+  return NMTUtil::flag_to_enum_name(f);
+}
+
+class MemFlagBitmap {
+  uint32_t _v;
+  STATIC_ASSERT(sizeof(_v) * BitsPerByte >= mt_number_of_types);
+
+public:
+  MemFlagBitmap(uint32_t v = 0) : _v(v) {}
+  MemFlagBitmap(const MemFlagBitmap& o) : _v(o._v) {}
+
+  uint32_t raw_value() const { return _v; }
+
+  void set_flag(MEMFLAGS f) {
+    const int bitno = (int)f;
+    _v |= nth_bit(bitno);
+  }
+
+  bool has_flag(MEMFLAGS f) const {
+    const int bitno = (int)f;
+    return _v & nth_bit(bitno);
+  }
+
+  bool has_any() const { return _v > 0; }
+
+};
+
+/// NMT virtual memory
```

Comment:
Not allocation.hpp, but lets put it into an own header. Arguable the NMT flags should be removed from allocation.hpp and moved somewhere else, since often I find myself including allocation.hpp just for the sake of mt_number_of_flags or some minor thing.